NÄDAL: 7 / MEESKOND: URBANSTYLE MARKETING DATA / ROLL: A 

1. Impordin pandase.

In [1]:
import pandas as pd

2. Laen sales.csv faili DataFrame'i df_sales.

In [2]:
df_sales = pd.read_csv("sales.csv")

3. Prindin shape ja head() tulemused.

In [3]:
print(df_sales.shape)
print(df_sales.head())

(15234, 11)
   sale_id        invoice_id   sale_date  customer_id  product_id  quantity  \
0        1  INV-202301-00001  2023-01-10       2588.0        1274         2   
1        2  INV-202301-00002  2023-01-16       4338.0        1207         2   
2        3  INV-202301-00003  2023-01-05       4673.0        1264         1   
3        4  INV-202301-00004  2023-01-02       4677.0        1341         3   
4        5  INV-202301-00005  2023-01-13       2390.0        1284         1   

   unit_price  total_price channel store_location payment_method  
0      234.79       469.58    pood        Tallinn          kaart  
1      241.13       482.26    pood          Pärnu      järelmaks  
2      258.46       221.19    pood          Pärnu      järelmaks  
3       45.21       135.63    pood          Tartu       sularaha  
4       99.57        99.57    pood          Tartu          kaart  


4. Laen customers.csv faili DataFrame'i df_customers.

In [4]:
df_customers = pd.read_csv("customers.csv")

5. Prindin shape ja head() tulemused.

In [5]:
print(df_customers.shape)
print(df_customers.head())

(3150, 9)
   customer_id first_name last_name                   email           phone  \
0         2001        Eha       Aas        eha.aas@telia.ee  +372 8713 1455   
1         2002      Aivar      Kõiv  aivar.koiv@outlook.com  +372 8943 8684   
2         2003      Maris    Rebane   maris.rebane@telia.ee  +372 5918 5726   
3         2004       Jaak    Talvik     jaak.talvik@mail.ee  +372 8554 4232   
4         2005      Raivo    Koppel  raivo.koppel@yahoo.com  +372 5298 4365   

       city registration_date loyalty_tier  birth_year  
0   Tallinn        2024-02-27          NaN        1973  
1  Haapsalu        2025-01-09       bronze        1988  
2     Tartu        2021-02-03          NaN        1999  
3   Tallinn        2023-11-12       silver        1974  
4   Tallinn        2023-05-22       bronze        2004  


5. Liidan sales ja customers tabelid customer_id alusel.

In [6]:
df = pd.merge(df_sales, df_customers, on='customer_id', how='left')

6. Prindin liidetud DataFrame'i shape, dtypes ja head() tulemused.

In [7]:
print(df.shape)
print(df.dtypes)
print(df.head())

(15234, 19)
sale_id                int64
invoice_id               str
sale_date                str
customer_id          float64
product_id             int64
quantity               int64
unit_price           float64
total_price          float64
channel                  str
store_location           str
payment_method           str
first_name               str
last_name                str
email                    str
phone                    str
city                     str
registration_date        str
loyalty_tier             str
birth_year           float64
dtype: object
   sale_id        invoice_id   sale_date  customer_id  product_id  quantity  \
0        1  INV-202301-00001  2023-01-10       2588.0        1274         2   
1        2  INV-202301-00002  2023-01-16       4338.0        1207         2   
2        3  INV-202301-00003  2023-01-05       4673.0        1264         1   
3        4  INV-202301-00004  2023-01-02       4677.0        1341         3   
4        5  INV-202301-00005

7. Kontrollin, et vajalikud veerud customer_id, sale_date, total_price ja email on olemas.

In [8]:
print(df[['customer_id', 'sale_date', 'total_price', 'email']].head())

   customer_id   sale_date  total_price                 email
0       2588.0  2023-01-10       469.58                   NaN
1       4338.0  2023-01-16       482.26    merle.luik@mail.ee
2       4673.0  2023-01-05       221.19  liina.saar@gmail.com
3       4677.0  2023-01-02       135.63   aili.pihl@yahoo.com
4       2390.0  2023-01-13        99.57   triin.lill@telia.ee


NÄDAL: 7 / MEESKOND: URBANSTYLE MARKETING DATA / ROLL: B

1. Prindin esialgse shape´i.

In [9]:
print("Esialgne shape:", df.shape)

Esialgne shape: (15234, 19)


2. Kontrollin duplikaatide arvu.

In [10]:
print("Duplikaadid:", df.duplicated(subset=['invoice_id']).sum())

Duplikaadid: 5116


3. Eemaldan duplikaadid.

In [11]:
df = df.drop_duplicates(subset=['invoice_id'], keep='first')

4. Kontrollin pärast eemaldamist shape´i.

In [12]:
print("Pärast duplikaatide eemaldamist:", df.shape)

Pärast duplikaatide eemaldamist: (10118, 19)


5. Kontrollin NULL väärtusi.

In [13]:
print("NULL-id:\n", df.isnull().sum())

NULL-id:
 sale_id                 0
invoice_id              0
sale_date               0
customer_id           988
product_id              0
quantity                0
unit_price              0
total_price             0
channel                 0
store_location       3462
payment_method          0
first_name            988
last_name             988
email                1944
phone                 988
city                  988
registration_date     988
loyalty_tier         4660
birth_year            988
dtype: int64


6. Eemaldan read, kus customer_id, sale_date või total_price on NULL ehk puuduvad.

In [14]:
df = df.dropna(subset=['customer_id', 'sale_date', 'total_price'])

7. Kontrollin pärast NULL-ide eemaldamist shape´i.

In [15]:
print("Pärast NULL-ide eemaldamist:", df.shape)

Pärast NULL-ide eemaldamist: (9130, 19)


8. Muudan sale_date tekstist kuupäevaks ja ühtlustan kuupäevade vormingu. 

In [16]:
df['sale_date'] = pd.to_datetime(df['sale_date'], format='mixed', dayfirst=True)

9. Eemaldan andmed alates 01.03.2025 (puudulikud andmed).

In [17]:
df = df[df['sale_date'] <= '2025-02-28']

10. Kontrollin, kas kuupäevavahemik on õige.

In [18]:
print("Kuupäevavahemik:", df['sale_date'].min(), "kuni", df['sale_date'].max())

Kuupäevavahemik: 2023-01-01 00:00:00 kuni 2025-02-28 00:00:00


11. Kontrollin pärast kuupäevade eemaldamist shape´i.

In [19]:
print("Pärast kuupäevade eemaldamist:", df.shape)

Pärast kuupäevade eemaldamist: (8887, 19)


12. Kontrollin, kas on negatiivseid total_price väärtusi.

In [20]:
print("Negatiivsed total_price väärtused:", (df['total_price'] < 0).sum())

Negatiivsed total_price väärtused: 175


13. Eemaldan negatiivsed total_price väärtused.

In [21]:
df = df[df['total_price'] > 0]

14. Kontrollin pärast negatiivsete total_price väärtuste eemaldamist shape´i.

In [22]:
print("Pärast negatiivsete väärtuste eemaldamist:", df.shape)

Pärast negatiivsete väärtuste eemaldamist: (8712, 19)


15. Puhastusraport: lõplik shape, unikaalsete klientide arv, kuupäevavahemik (min kuni max).

In [23]:
print("Lõplik shape:", df.shape)
print("Unikaalseid kliente:", df['customer_id'].nunique())
print("Kuupäevavahemik:", df['sale_date'].min(), "kuni", df['sale_date'].max())

Lõplik shape: (8712, 19)
Unikaalseid kliente: 2515
Kuupäevavahemik: 2023-01-01 00:00:00 kuni 2025-02-28 00:00:00


NÄDAL: 7 / MEESKOND: URBANSTYLE MARKETING DATA / ROLL: C

1. Määran viitekuupäeva, kui kaua on kliendi viimasest ostust möödas.

In [24]:
today = pd.to_datetime('2025-02-28')

2. Leian iga kliendi viimase ostu kuupäeva.

In [25]:
last_purchase = df[df['sale_date'] <= today].groupby('customer_id')['sale_date'].max()

3. Arvutan Recency (viitekuupäev − kliendi viimane ost = mitu päeva on viimasest ostust möödas).

In [26]:
recency = (today - last_purchase).dt.days

4. Arvutan Frequency (iga kliendi ostude arv).

In [27]:
frequency = df.groupby('customer_id')['sale_id'].count()

5. Arvutan Monetary (iga kliendi kogukulutus).

In [28]:
monetary = df.groupby('customer_id')['total_price'].sum()

6. Ühendan recency, frequency ja monetary. 

In [29]:
rfm = pd.merge(recency, frequency, on='customer_id')
rfm = pd.merge(rfm, monetary, on='customer_id')

rfm.columns = ['Recency', 'Frequency', 'Monetary']

7. Määran Recency skoori (hiljuti ostnud klient saab kõrgema skoori).

In [30]:
rfm['R_score'] = pd.qcut(rfm['Recency'], 5, labels=[5, 4, 3, 2, 1])

8. Paljudel klientidel kordub täpselt sama ostude arv, seega kasutan rank()-i. 


In [31]:
frequency_rank = rfm['Frequency'].rank(method='first')

9. Määran Frequency skoori.

In [32]:
rfm['F_score'] = pd.qcut(frequency_rank, 5, labels=[1, 2, 3, 4, 5])

10. Määran Monetary skoori.

In [33]:
rfm['M_score'] = pd.qcut(rfm['Monetary'], 5, labels=[1, 2, 3, 4, 5])

11. Arvutan RFM koguskoori.

In [34]:
rfm['RFM_Score'] = (
    rfm['R_score'].astype(int)
    + rfm['F_score'].astype(int)
    + rfm['M_score'].astype(int)
)

12. Kontrollin, millised skoorid tekkisid.

In [35]:
print(rfm['RFM_Score'].value_counts().sort_index())

RFM_Score
3     124
4     129
5     186
6     197
7     270
8     231
9     239
10    265
11    231
12    188
13    180
14    160
15    115
Name: count, dtype: int64


13. Loon segmendid (13-15 = VIP Champions, 10-12 = Loyal, 7-9 = Potential, 4-6 = At Risk, 3 = Lost). 

In [36]:
def segment(score):
    if score >= 13:
        return 'VIP Champions'
    elif score >= 10:
        return 'Loyal'
    elif score >= 7:
        return 'Potential'
    elif score >= 4:
        return 'At Risk'
    else:
        return 'Lost'

rfm['Segment'] = rfm['RFM_Score'].apply(segment)

14. Segmentide kokkuvõte.

In [37]:
print(rfm['Segment'].value_counts().sort_index())

Segment
At Risk          512
Lost             124
Loyal            684
Potential        740
VIP Champions    455
Name: count, dtype: int64


NÄDAL: 7 / MEESKOND: URBANSTYLE MARKETING DATA / ROLL: D

1. Impordin Plotly.

In [38]:
import plotly.express as px

2. Diagrammi 1 - segmentide jaotus.

In [39]:
segment_counts = rfm['Segment'].value_counts().reset_index()
segment_counts.columns = ['Segment', 'Klientide arv']

fig = px.bar(
    segment_counts,
    x='Segment',
    y='Klientide arv',
    title='UrbanStyle kliendisegmentide jaotus',
    labels={
        'Segment': 'Kliendisegment',
        'Klientide arv': 'Klientide arv'
    }
)

fig.show()

3. Diagramm 2 - viimane ost vs kogukulutus.

In [40]:
fig = px.scatter(
    rfm,
    x='Recency',
    y='Monetary',
    color='Segment',
    size='Frequency',
    title='UrbanStyle kliendisegmendid (RFM)',
    labels={
        'Recency': 'Päevi viimasest ostust',
        'Monetary': 'Kogukulutus (EUR)',
        'Frequency': 'Ostude arv',
        'Segment': 'Kliendisegment'
    }
)

fig.show()

4. Diagramm 3 — TOP 10 VIP klienti kogukulutuse järgi.

In [41]:
top_vip = (
    rfm[rfm['Segment'] == 'VIP Champions']
    .nlargest(10, 'Monetary')
)

fig = px.bar(
    top_vip,
    x=top_vip.index,
    y='Monetary',
    title='Top 10 VIP klienti kogukulu järgi',
    labels={
        'x': 'Kliendi ID',
        'Monetary': 'Kogukulu'
    }
)

fig.update_xaxes(title_text='Kliendi ID')

fig.show()

5. Leian VIP klientide osakaalu kogu RFM Monetary summast. Kontrollin VIP ja At Risk segmendi klientide arvu.

In [42]:
vip = rfm[rfm['Segment'] == 'VIP Champions']

vip_share = vip['Monetary'].sum() / rfm['Monetary'].sum() * 100

print(f"VIP klientide käibeosakaal: {vip_share:.1f}%")


print(f"VIP klientide arv: {len(vip)}")


at_risk = rfm[rfm['Segment'] == 'At Risk']

print(f"At Risk klientide arv: {len(at_risk)}")

VIP klientide käibeosakaal: 43.6%
VIP klientide arv: 455
At Risk klientide arv: 512


Äritõlgendus: 455 VIP Champions klienti moodustavad 43,6% klientide käibest. 512 klienti kuulub At Risk segmenti, mis näitab võimalust kliente uuesti aktiveerida. Fookuses võiks olla VIP klientide hoidmine ja At Risk klientide tagasivõitmine.

SOOVITUSED
- VIP programm: pakkuda VIP klientidele eritingimusi ja personaalseid pakkumisi.
- Win-back kampaania: suunata At Risk klientidele sihitud pakkumisi, et neid uuesti ostma aktiveerida.
- Nurture programm: kaasata uusi kliente järjepideva ja personaalse kommunikatsiooniga.

Salvestan segmendid CSV failina rfm_segments.csv 

In [43]:
rfm.to_csv('rfm_segments.csv', index=False)